### PR TITLE
[fenecon] Initial contribution

### LINKED ISSUE #17173: [fenecon] New binding for FENECON
## Goal

New binding for [FENECON energy storage system](https://fenecon.de/).

## Motivation
The [FENECON energy storage system](https://fenecon.de/) device provides a [REST-API](https://docs.fenecon.de/_/de/fems/fems-app/OEM_App_REST_JSON.html) which makes it possible to request status information from FENECON system to make decisions for home automation based on the current energy management.

### PR DESCRIPTION
**Pull request by nixoso**
- Binding for FENECON via REST-API access implemented
- Documentation for FENECON added
- The implementation Closes #17173

### CODE REVIEW

File: bundles/org.openhab.binding.fenecon/README.md
Comment (lsiepel) [inline bundles/org.openhab.binding.fenecon/README.md:38]: Please adjust all channel id's according to the naming convention `lower-case-hyphen`. Readme, thing-types.xml, binding constants and any other place these are used.
For reference: https://www.openhab.org/docs/developer/guidelines.html#naming-convention

File: bundles/org.openhab.binding.fenecon/README.md
Comment (lsiepel) [inline bundles/org.openhab.binding.fenecon/README.md:104]: You can consider to create a tabstrip with `:::` even if it has one tab, it can show DSL, other tabs might be added later for JS, Python or other.

File: bundles/org.openhab.binding.fenecon/pom.xml
Comment (lsiepel) [inline bundles/org.openhab.binding.fenecon/pom.xml:16]: Please add connection tag and if possible others too.

File: bundles/org.openhab.binding.fenecon/README.md
Comment (lsiepel) [inline bundles/org.openhab.binding.fenecon/README.md:11]: Is a `Thing` with the same name as the binding future proof?  Maybe better to change to `device` or something else. In the future more `Things` might be added.

File: bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml
Comment (lsiepel) [inline bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml:40]: Double check: guest password that is provided by default should give enough autorisations for the binding to work normally? If not it should not be marked as advanced.

File: bundles/org.openhab.binding.fenecon/src/main/java/org/openhab/binding/fenecon/internal/FeneconHandler.java
Comment (lsiepel) [inline bundles/org.openhab.binding.fenecon/src/main/java/org/openhab/binding/fenecon/internal/FeneconHandler.java:64]: Please extract these strings as constants into `FeneconBindingConstants.java`

File: bundles/org.openhab.binding.fenecon/src/main/java/org/openhab/binding/fenecon/internal/FeneconHandler.java
Comment (lsiepel) [inline bundles/org.openhab.binding.fenecon/src/main/java/org/openhab/binding/fenecon/internal/FeneconHandler.java:80]:
```suggestion
        logger.debug("FENECON: initialize REST-API connection to {} with polling interval: {} sec", getBaseUrl(config),
```

File: bundles/org.openhab.binding.fenecon/src/main/java/org/openhab/binding/fenecon/internal/FeneconHandler.java
Comment (lsiepel) [inline bundles/org.openhab.binding.fenecon/src/main/java/org/openhab/binding/fenecon/internal/FeneconHandler.java:86]:
```suggestion
        pollingJob = scheduler.scheduleWithFixedDelay(this::pollingCode, 0, config.refreshInterval, TimeUnit.SECONDS);
```

Is tehre a need to wait 5 seconds?

File: bundles/org.openhab.binding.fenecon/src/main/java/org/openhab/binding/fenecon/internal/FeneconHandler.java
Comment (lsiepel) [inline bundles/org.openhab.binding.fenecon/src/main/java/org/openhab/binding/fenecon/internal/FeneconHandler.java:102]: For good seperation of concerns it would be better to extract these and other http/api handling out of this `FeneconHandler` class and move it into it's own `ApiController` or whatever you name it. 

A common approach is to call public methods on that `ApiController` and whenever it fails, throw custom exceptions within that `ApiController`  class. Those can be catched by the calling `FeneconHandler` class that can set the `Thing` state accordingly. 

An example is the PlugwiseHA binding, maybe that can be used as some kind of inspiration.

File: bundles/org.openhab.binding.fenecon/src/main/java/org/openhab/binding/fenecon/internal/FeneconHandler.java
Comment (lsiepel) [inline bundles/org.openhab.binding.fenecon/src/main/java/org/openhab/binding/fenecon/internal/FeneconHandler.java:160]: See other comment, use constants for the channel id's

File: bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/i18n/fenecon.properties
Comment (lsiepel) [inline bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/i18n/fenecon.properties:1]: Please create contents for this file with: `mvn i18n:generate-default-translations`

File: bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml
Comment (lsiepel) [inline bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml:62]: You can consider to add these values with an option tag. as it is readonly it cannot be set, but when used with option tags it can be translated.

File: bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml
Comment (lsiepel) [inline bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml:66]: See other comment, about naming convention, they also apply to channel type id's

File: bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml
Comment (lsiepel) [inline bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml:78]:
```suggestion
		<state readOnly="true" pattern="%.0f %%"/>
```

File: bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml
Comment (lsiepel) [inline bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml:97]:
```suggestion
		<description>Indicates if there is no power from the grid and the emergency power mode is on.</description>
```

File: bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml
Comment (lsiepel) [inline bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml:138]:
```suggestion
		<label>Bought Grid Energy</label>
```

Comment (lsiepel) [reply bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml:138]: More common is import and or export. I'm not a native english speaker, but these labels might need some improvement. Maybe look at the Tibber binding that has similar channels.

**Review by lsiepel — CHANGES_REQUESTED** [review]
Thank you for this contribution. Don't forget to add yourself to the codeowner file. After all comments have been addressed i have to do another review round.
Please also check compilation and SAT warnings (not looked at that right now)

File: bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml
Comment (nixoso) [reply bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml:40]: Yes, the guest password is enough, so it is specified in the [official documentation](https://docs.fenecon.de/_/de/fems/fems-app/OEM_App_REST_JSON.html)

File: bundles/org.openhab.binding.fenecon/README.md
Comment (nixoso) [reply bundles/org.openhab.binding.fenecon/README.md:104]: Do you have an example with the tabstrip, unfortunately I can't get it to work?

File: bundles/org.openhab.binding.fenecon/pom.xml
Comment (nixoso) [reply bundles/org.openhab.binding.fenecon/pom.xml:16]: Done

File: bundles/org.openhab.binding.fenecon/src/main/java/org/openhab/binding/fenecon/internal/FeneconHandler.java
Comment (nixoso) [reply bundles/org.openhab.binding.fenecon/src/main/java/org/openhab/binding/fenecon/internal/FeneconHandler.java:86]: No, that was still from testing, I changed it.

File: bundles/org.openhab.binding.fenecon/src/main/java/org/openhab/binding/fenecon/internal/FeneconHandler.java
Comment (nixoso) [reply bundles/org.openhab.binding.fenecon/src/main/java/org/openhab/binding/fenecon/internal/FeneconHandler.java:80]: Done

File: bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml
Comment (nixoso) [reply bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml:78]: I do not understand this, everywhere in other bindings %unit% is explicitly specified for percent. I have now added a unitHint.

File: bundles/org.openhab.binding.fenecon/README.md
Comment (nixoso) [reply bundles/org.openhab.binding.fenecon/README.md:38]: Done

File: bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml
Comment (nixoso) [reply bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml:62]: Done with the naming convention fix.

File: bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml
Comment (nixoso) [reply bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml:97]: Done

File: bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml
Comment (nixoso) [reply bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml:138]: Done

File: bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml
Comment (nixoso) [reply bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml:66]: Done

File: bundles/org.openhab.binding.fenecon/README.md
Comment (nixoso) [reply bundles/org.openhab.binding.fenecon/README.md:11]: Good point, I changed it to `rest-api`.

File: bundles/org.openhab.binding.fenecon/src/main/java/org/openhab/binding/fenecon/internal/FeneconHandler.java
Comment (nixoso) [reply bundles/org.openhab.binding.fenecon/src/main/java/org/openhab/binding/fenecon/internal/FeneconHandler.java]: Done

File: bundles/org.openhab.binding.fenecon/src/main/java/org/openhab/binding/fenecon/internal/FeneconHandler.java
Comment (nixoso) [reply bundles/org.openhab.binding.fenecon/src/main/java/org/openhab/binding/fenecon/internal/FeneconHandler.java]: Done

File: bundles/org.openhab.binding.fenecon/src/main/java/org/openhab/binding/fenecon/internal/FeneconHandler.java
Comment (nixoso) [reply bundles/org.openhab.binding.fenecon/src/main/java/org/openhab/binding/fenecon/internal/FeneconHandler.java]: @lsiepel So now everything should be done review-wise 😊

File: bundles/org.openhab.binding.fenecon/README.md
Comment (lsiepel) [reply bundles/org.openhab.binding.fenecon/README.md]: But what if the api adds more devices that need some configuration? The current line up exists around home 10 devices (if I’m correct, just had a Quick look at their website) so maybe change it to `home-device` ?

File: bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml
Comment (lsiepel) [reply bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml:85]: Percentage is a little special compared to others. AFAIK the ui hint should also work.

**Comment by lsiepel** [comment]
There is a conflict to fix. I’ll come back to this. Two weeks AFK.

File: bundles/org.openhab.binding.fenecon/README.md
Comment (lsiepel) [reply bundles/org.openhab.binding.fenecon/README.md:104]: Check the pushover readme here https://github.com/openhab/openhab-addons/blob/8c29b5973e5fa1159f11b325eb3cd6723d690176/bundles/org.openhab.binding.pushover/README.md?plain=1#L118

Comment (nixoso) [reply bundles/org.openhab.binding.fenecon/README.md:104]: Thank you, I needed that example 😊

File: bundles/org.openhab.binding.fenecon/README.md
Comment (nixoso) [reply bundles/org.openhab.binding.fenecon/README.md]: Yes, `home-device` is the better choice, I have changed it.

File: bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml
Comment (lsiepel) [reply bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml:90]: %unit% together with the unitHint seems to be the best solution.

**Comment by nixoso** [comment]
@lsiepel can you please take a look at the _"PR-openHAB addons"_ because of the errors (https://github.com/openhab/openhab-addons/pull/17174/commits/de826404e373ea2b97b03fd6680c0d7ee34ea652, https://github.com/openhab/openhab-addons/pull/17174/commits/c7af0d99bffd6d3e769a21c38d042257f2fab699) in the build, unfortunately I can't find out anything about it and wouldn't see it in relation to my changes.

**Comment by lsiepel** [comment]
Well there is a conflict and the bundles/pom.xml is showing many ampty lines, maybe you want to revert that file and add your binding again. I will try look at the other changes somewhere this week.

**Comment by nixoso** [comment]
@lsiepel 
I could finally solve the problem with the formatting. But now there are still problems with the “PR-openHAB-Addons” but the error does not seem to be caused by the change. Do you have another tip?

**Comment by lsiepel** [comment]
> @lsiepel I could finally solve the problem with the formatting. But now there are still problems with the “PR-openHAB-Addons” but the error does not seem to be caused by the change. Do you have another tip?

Ignore that for now. Jenkins is somewhat unstable

File: bundles/org.openhab.binding.fenecon/pom.xml
Comment (lsiepel) [reply bundles/org.openhab.binding.fenecon/pom.xml:16]: The connection tag is missing and the inceptionYear/developers tags are not used for official openHAB bindings (they are only for marketplace/3rd party), please remove those.

File: bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml
Comment (lsiepel) [inline bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml:66]:
```suggestion
		<label>System State</label>
```

File: bundles/org.openhab.binding.fenecon/README.md
Comment (lsiepel) [inline bundles/org.openhab.binding.fenecon/README.md:99]:
```suggestion
```

File: bundles/org.openhab.binding.fenecon/README.md
Comment (lsiepel) [inline bundles/org.openhab.binding.fenecon/README.md:55]:
```suggestion
```

File: bundles/org.openhab.binding.fenecon/README.md
Comment (lsiepel) [inline bundles/org.openhab.binding.fenecon/README.md:30]:
```suggestion
```

File: bundles/org.openhab.binding.fenecon/src/main/java/org/openhab/binding/fenecon/internal/api/FeneconController.java
Comment (lsiepel) [inline bundles/org.openhab.binding.fenecon/src/main/java/org/openhab/binding/fenecon/internal/api/FeneconController.java:59]: A specific reason for not using the common httpclient ?

**Review by lsiepel — CHANGES_REQUESTED** [review]
Some cleanup suggestions and a question. I think we are almost there.

File: bundles/org.openhab.binding.fenecon/README.md
Comment (nixoso) [reply bundles/org.openhab.binding.fenecon/README.md]: Done

File: bundles/org.openhab.binding.fenecon/README.md
Comment (nixoso) [reply bundles/org.openhab.binding.fenecon/README.md]: Done

File: bundles/org.openhab.binding.fenecon/README.md
Comment (nixoso) [reply bundles/org.openhab.binding.fenecon/README.md]: Done

File: bundles/org.openhab.binding.fenecon/src/main/java/org/openhab/binding/fenecon/internal/api/FeneconController.java
Comment (nixoso) [reply bundles/org.openhab.binding.fenecon/src/main/java/org/openhab/binding/fenecon/internal/api/FeneconController.java:59]: I always think it's good to integrate as few external libs as possible, especially when Java provides solutions for this.  In this case, the Java client is perfectly suitable and automatically receives security fixes via the Java update.

File: bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml
Comment (nixoso) [reply bundles/org.openhab.binding.fenecon/src/main/resources/OH-INF/thing/thing-types.xml]: I have removed the unnecessary tags from the pom.xml. However, the connection tag is included in the addon.xml. Or do you mean something else?

File: bundles/org.openhab.binding.fenecon/pom.xml
Comment (lsiepel) [inline bundles/org.openhab.binding.fenecon/pom.xml:16]: See https://github.com/openhab/openhab-addons/pull/17174#discussion_r1732757206 the connection tag is still missing

**Review by lsiepel — CHANGES_REQUESTED** [review]
Two existing comments are left.

File: bundles/org.openhab.binding.fenecon/src/main/java/org/openhab/binding/fenecon/internal/api/FeneconController.java
Comment (lsiepel) [reply bundles/org.openhab.binding.fenecon/src/main/java/org/openhab/binding/fenecon/internal/api/FeneconController.java:59]: Perfectly sane 'rule'. For openHAB we like to control threads, resources and have made some design choices and that include the usage of the common httpclient unless there is some binding specific condition that can not be met with the common client.
So it would be nice if you inject the client and re-use that.

File: bundles/org.openhab.binding.fenecon/pom.xml
Comment (nixoso) [reply bundles/org.openhab.binding.fenecon/pom.xml:16]: I have added the connection tag in the addon.xml. I can't find a connection tag in the pom.xml in any other bundle. Can you link me to an example?

Comment (lsiepel) [reply bundles/org.openhab.binding.fenecon/pom.xml:16]: Sorry, wrong comment on wrong file. Please ignore. :-/

File: bundles/org.openhab.binding.fenecon/src/main/java/org/openhab/binding/fenecon/internal/api/FeneconController.java
Comment (nixoso) [reply bundles/org.openhab.binding.fenecon/src/main/java/org/openhab/binding/fenecon/internal/api/FeneconController.java]: I have replaced the Java client with the Jetty http client.

**Review by lsiepel — APPROVED** [review]
Thanks! LGTM.

Now, you could add your binding's logo to the openHAB website. See https://next.openhab.org/docs/developer/addons/#add-your-add-on-s-logo-to-the-openhab-website-and-the-ui

**Pull request merged**
